### PR TITLE
fix(core): bundling directory access permission is too restrictive

### LINKED ISSUE #8757: [aws-lambda-nodejs] Broken in BitBucket Pipelines
There seems to be a problem with NodejsFunction in Bitbucket Pipelines. When I attempt to create a lambda using that construct and synthesize, I get an error about `lscpu` and another `Error: EACCES: permission denied, open '/asset-output/index.js'`.

### Reproduction Steps
**bug-stack.ts**
```ts
import * as cdk from '@aws-cdk/core';
import * as nodejs from '@aws-cdk/aws-lambda-nodejs';

export class BugStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    new nodejs.NodejsFunction(this, 'handler');
  }
}
```
**bug-stack.handler.ts**
```ts
export const handler = async () => {
  console.log("Hello, world!");
};
```
**bitbucket-pipelines.yml**
```yml
image: node:12
pipelines:
  default:
    - step:
        caches:
          - node
          - docker
        services:
          - docker
        script:
          - node -v
          - yarn
          - yarn cdk --version
          - yarn cdk synth
        after-script:
          - ls -la
          - ls -la .cdk.staging
```

### Error Log
```
yarn run v1.22.4
warning package.json: No license field
$ cdk synth
Failed to run bundling Docker image for asset BugStack/handler/Code/Stage: Error: [Status 1] stdout: 
stderr: /bin/sh: lscpu: not found
events.js:292
      throw er; // Unhandled 'error' event
      ^
Error: EACCES: permission denied, open '/asset-output/index.js'
Emitted 'error' event on WriteStream instance at:
    at internal/fs/streams.js:376:12
    at FSReqCallback.oncomplete (fs.js:155:23) {
  errno: -13,
  code: 'EACCES',
  syscall: 'open',
  path: '/asset-output/index.js'
}
Subprocess exited with error 1
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

### Environment

  - **CLI Version      :** 1.47.0
  - **Framework Version:** 1.47.0
  - **Node.js Version:** 
  - **OS               :** BitBucket Pipelines, based on the node:12 image
 - **Language (Version):**  TypeScript (3.7.2)

### Other
This might be related to #8707, #8492 

Here's what I get from ls -la in my pipeline:
```
+ ls -la
total 548
drwxrwxrwx.   9 root root   4096 Jun 26 17:20 .
drwxr-xr-x.   8 root root   4096 Jun 26 17:18 ..
drwxrwxrwx.   3 root root   4096 Jun 26 17:20 .cdk.staging
drwxrwxrwx.   8 root root   4096 Jun 26 17:18 .git
-rw-rw-rw-.   1 root root    135 Jun 26 17:18 .gitignore
-rw-rw-rw-.   1 root root     65 Jun 26 17:18 .npmignore
-rw-rw-rw-.   1 root root    543 Jun 26 17:18 README.md
drwxrwxrwx.   2 root root   4096 Jun 26 17:18 bin
-rw-rw-rw-.   1 root root    322 Jun 26 17:18 bitbucket-pipelines.yml
-rw-rw-rw-.   1 root root    152 Jun 26 17:18 cdk.json
drwxrwxrwx.   2 root root   4096 Jun 26 17:20 cdk.out
-rw-rw-rw-.   1 root root    130 Jun 26 17:18 jest.config.js
drwxrwxrwx.   3 root root   4096 Jun 26 17:20 lib
drwxr-xr-x. 470 root root  20480 Jun 26 17:19 node_modules
-rw-rw-rw-.   1 root root 277830 Jun 26 17:18 package-lock.json
-rw-rw-rw-.   1 root root    604 Jun 26 17:18 package.json
drwxrwxrwx.   2 root root   4096 Jun 26 17:18 test
-rw-rw-rw-.   1 root root    598 Jun 26 17:18 tsconfig.json
-rw-rw-rw-.   1 root root 195681 Jun 26 17:18 yarn.lock
```

In .cdk.staging.
```
+ ls -la .cdk.staging
total 12
drwxrwxrwx. 3 root root 4096 Jun 26 17:20 .
drwxrwxrwx. 9 root root 4096 Jun 26 17:20 ..
drwx------. 2 root root 4096 Jun 26 17:20 asset-bundle-gYtxGP
```

If I change the pipeline user to 1000, by changing the yml image to `image: { name: node:12, run-as-user: 1000 }`, the synth still fails with the same error. The only difference I can see is that the .cdk.staging and cdk.out directories have a different owner. (1000, which is the 'node' user)

--- 

This is :bug: Bug Report

### PR DESCRIPTION
**Pull request by misterjoshua**
The new bundler uses `mkdtempSync` to pre-create uniquely named directories for asset staging. But, `mkdtempSync` creates the staging directories with a restrictive `0700 & ~umask` mode, rather than `mkdir`'s usual `0777 & ~umask` mode.

In Bitbucket Pipelines, these restrictive permissions prevent the bundler from accessing its `/asset-output` volume. And, if the bundler can't access `/asset-output`, bundling fails.

This fix chmods the asset staging directory to 0777. This change fixes my Bitbucket Pipelines issue.

Closes #8757 

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by eladb** [comment]
@jogold  can you take a look at this please?

File: packages/@aws-cdk/core/lib/asset-staging.ts [hunk 1]
```diff
@@ -144,6 +144,9 @@ export class AssetStaging extends Construct {
 
     // Create temp directory for bundling inside the temp staging directory
     const bundleDir = path.resolve(fs.mkdtempSync(path.join(stagingTmp, 'asset-bundle-')));
+    // Chmod the bundleDir to full access after applying the process the umask.
+    // tslint:disable-next-line:no-bitwise
+    fs.chmodSync(bundleDir, 0o777 & (~process.umask()));
```
Comment (jogold) [inline packages/@aws-cdk/core/lib/asset-staging.ts:149]: Looks like calling `process.umask()` without no argument is now deprecated in Node.js v14: https://nodejs.org/docs/latest-v14.x/api/process.html#process_process_umask

What would be the alternative? What about using only `Oo777`, would it be OK?

Comment (misterjoshua) [reply packages/@aws-cdk/core/lib/asset-staging.ts:149]: Another alternative is to set a new umask (to get the return value), then restore the original umask. I've seen a [getumask manpage](https://man7.org/linux/man-pages/man3/getumask.3.html) that describes this approach. But, `0o777` would certainly satisfy my use case. I can't think of any major downsides. What's your preference?

Comment (jogold) [reply packages/@aws-cdk/core/lib/asset-staging.ts:149]: > But, mode `0o777` alone would certainly satisfy my use case. I can't think of any major downsides.

OK, let's go for this.

**Comment by jogold** [comment]
@eladb LGTM, maybe change the PR title to something like `fix(core): bundling directory access permission is too restrictive`

**Pull request merged**